psi4/driver/procrouting/diis.py

```diff
@@ -0,0 +1,212 @@
+from enum import Enum
+from psi4 import core
+from psi4.driver import psifiles as psif
+
+import numpy as np
+
+class RemovalPolicy(Enum):
+    LargestError = 1
+    OldestAdded = 2
+
+class StoragePolicy(Enum):
+    InCore = 1
+    OnDisk = 2
+
+def axpy(y, alpha, x):
+    if isinstance(y, (core.Matrix, core.Vector)):
+        y.axpy(alpha, x)
+    elif isinstance(y, (core.dpdbuf4, core.dpdfile2)):
+        y.axpy_matrix(x, alpha)
+    else:
+        raise TypeError("Unrecognized object type for DIIS.")
+
+def template_helper(template, *args):
+    if template:
+        raise Exception("Template already set.")
+
+    for arg in args:
+        if isinstance(arg, core.Vector):
+            template.append([arg.dimpi()])
+        elif isinstance(arg, (core.Matrix, core.dpdfile2, core.dpdbuf4)):
+            template.append([arg.rowdim(), arg.coldim()])
+        else:
+            raise TypeError("Unrecognized object type for DIIS.")
+
+class DIIS:
+
+    def __init__(self, max_vecs: int, name: str, removal_policy = RemovalPolicy.LargestError, storage_policy = StoragePolicy.OnDisk):
+        assert isinstance(removal_policy, RemovalPolicy)
```

Comment:
generally `assert` can be optimized away when python code is compiled to bytecode with optimization [python3 docs](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement).

```
The current code generator emits no code for an assert statement when optimization is requested at compile time.
```
one can replace all occurrences with `if not` + runtime exception, but it's less concise. 

I guess it's a question is how defensive psi4 is supposed to be wrt to that runtime type checking vs the readability of `if not` clauses.

Reply:
`assert` isn't exactly user-friendly, so I might as well replace them.